File: src/commands/drive.ts
// ⚠️ LEGACY TYPESCRIPT PORT — NOT the live `mm` binary. The live CLI is Go;
// this command lives in internal/cmd/. Editing this file changes nothing in
// `mm` (it only builds the separate, unused `mm-ts`). Fix the .go file instead.

/**
 * mm drive — light Drive surface: list, doc-from-markdown, rename/move.
 *
 *   mm drive ls [--q "name contains 'invoice'"] [--max 25]
 *
 *   mm drive doc <name> --file path/to/notes.md
 *   mm drive doc <name> < notes.md               # stdin works too
 *   mm drive doc <name> --text "Plain string body" --mime text/plain
 *
 *   mm drive mv <id> --name "New name"
 *   mm drive mv <id> --parent <folder-id> [--unparent <old-folder-id>]
 *
 * `<name>` is what the Doc will be called in Drive. The content is
 * converted by Drive on import — markdown becomes formatted Doc
 * (headings, lists, code blocks), text/html becomes the same, plain
 * text becomes plain text.
 */
import { readFileSync } from 'node:fs';
import { hubApi } from '../hub';
import type {
	HubDriveListResp,
	HubDriveCreateDocResp,
	HubDriveUpdateResp,
} from '../wire';

export async function driveDispatch(
	command: string,
	args: string[],
	flags: { json?: boolean }
) {
	const json = flags?.json || false;
	switch (command) {
		case '':
		case 'help':
		case '--help':
		case '-h':
			printDriveHelp();
			return;
		case 'ls':
		case 'list':
			return driveList(args, json);
		case 'doc':
			return driveDoc(args, json);
		case 'rename':
		case 'mv':
		case 'move':
			return driveMove(args, json);
		default:
			console.error(`Unknown command: mm drive ${command}`);
			console.error('Try `mm drive help`.');
			process.exit(1);
	}
}

export function printDriveHelp() {
	console.log(`mm drive — Google Drive

Subcommands:
  ls | list                List files (filter with --q)
  doc <file.md>            Create a Google Doc from a markdown file
  mv | rename | move <id>  Rename and/or move a file

Examples:
  mm drive mv <id> --name "New name"
  mm drive mv <id> --parent <folder-id>
  mm drive mv <id> --parent <new> --unparent <old>      # move
  mm drive mv <id> --name "X" --parent <new> --unparent <old>

Flags:
  --q "<query>"            Search query (Drive search syntax)
  --max <n>                Max results (default 20)
  --name "<name>"          New name for mv
  --parent <folder-id>     Add a parent (move into)
  --unparent <folder-id>   Remove a parent (move out of)
  --folder <id>            Target folder for doc creation
  --account <slug|email>   Pick a linked Google account
  --json                   Parseable output`);
}

function parseFlags(args: string[]): { flags: Record<string, string>; positional: string[] } {
	const flags: Record<string, string> = {};
	const positional: string[] = [];
	for (let i = 0; i < args.length; i++) {
		const a = args[i];
		const eq = a.match(/^--([a-z][a-z-]*)=(.+)$/s);
		if (eq) {
			flags[eq[1]] = eq[2];
			continue;
		}
		const longFlag = a.match(/^--([a-z][a-z-]*)$/);
		if (longFlag) {
			const next = args[i + 1];
			if (next === undefined || next.startsWith('--')) {
				flags[longFlag[1]] = 'true';
			} else {
				flags[longFlag[1]] = next;
				i++;
			}
			continue;
		}
		positional.push(a);
	}
	return { flags, positional };
}

function fmtWhen(iso: string | undefined): string {
	if (!iso) return '—';
	const d = new Date(iso);
	if (Number.isNaN(d.getTime())) return iso;
	const diff = Date.now() - d.getTime();
	if (diff < 60_000) return 'just now';
	if (diff < 3_600_000) return `${Math.round(diff / 60_000)}m`;
	if (diff < 86_400_000) return `${Math.round(diff / 3_600_000)}h`;
	if (diff < 7 * 86_400_000) return `${Math.round(diff / 86_400_000)}d`;
	return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'short' });
}

function mimeLabel(m: string): string {
	if (m === 'application/vnd.google-apps.document') return 'doc';
	if (m === 'application/vnd.google-apps.spreadsheet') return 'sheet';
	if (m === 'application/vnd.google-apps.presentation') return 'slide';
	if (m === 'application/vnd.google-apps.folder') return 'folder';
	if (m.startsWith('image/')) return 'img';
	if (m === 'application/pdf') return 'pdf';
	return m.split('/').pop() ?? m;
}

async function driveList(args: string[], json: boolean) {
	const { flags } = parseFlags(args);
	const payload: Record<string, unknown> = {};
	if (flags.q) payload.q = flags.q;
	if (flags.max) payload.max = Number(flags.max);
	if (flags.account) payload.accountSlug = flags.account;

	const data = await hubApi<HubDriveListResp>('drive', 'list', payload);
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}
	if (data.files.length === 0) {
		console.log('No files match.');
		return;
	}
	for (const f of data.files) {
		const when = fmtWhen(f.modifiedTime).padStart(4);
		const kind = mimeLabel(f.mimeType).padEnd(6).slice(0, 6);
		console.log(`  ${when}  ${kind}  ${f.name}`);
		console.log(`        ${f.webViewLink ?? '(no link)'}`);
	}
}

async function readStdin(): Promise<string> {
	// Bun + Node compatible.
	const chunks: Buffer[] = [];
	for await (const chunk of process.stdin as unknown as AsyncIterable<Buffer | string>) {
		chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	}
	return Buffer.concat(chunks).toString('utf-8');
}

async function driveMove(args: string[], json: boolean) {
	const { flags, positional } = parseFlags(args);
	const fileId = positional[0];
	if (!fileId) {
		console.error('Usage: mm drive mv <id> [--name "X"] [--parent <folder-id>] [--unparent <folder-id>]');
		process.exit(1);
	}
	const payload: Record<string, unknown> = { fileId };
	if (flags.name) payload.name = flags.name;
	if (flags.parent) payload.addParents = [flags.parent];
	if (flags.unparent) payload.removeParents = [flags.unparent];
	if (flags.account) payload.accountSlug = flags.account;
	if (!payload.name && !payload.addParents && !payload.removeParents) {
		console.error('Nothing to do — pass at least one of --name, --parent, --unparent.');
		process.exit(1);
	}

	const data = await hubApi<HubDriveUpdateResp>('drive', 'update', payload);
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}
	console.log(`✓ ${data.name}`);
	if (data.webViewLink) console.log(`  ${data.webViewLink}`);
}

async function driveDoc(args: string[], json: boolean) {
	const { flags, positional } = parseFlags(args);
	const name = positional.join(' ').trim();
	if (!name) {
		console.error('Usage: mm drive doc <name> --file path.md  (or pipe in via stdin)');
		console.error('       mm drive doc <name> --text "literal body" [--mime text/plain|text/markdown|text/html]');
		process.exit(1);
	}

	// Pick the content source: --file > --text > stdin.
	let content: string;
	const sourceMime =
		(flags.mime as 'text/markdown' | 'text/plain' | 'text/html' | undefined) ?? 'text/markdown';
	if (flags.file) {
		try {
			content = readFileSync(flags.file, 'utf-8');
		} catch (err) {
			console.error(`Couldn't read --file ${flags.file}: ${(err as Error).message}`);
			process.exit(1);
		}
	} else if (flags.text) {
		content = flags.text;
	} else if (!process.stdin.isTTY) {
		content = await readStdin();
	} else {
		console.error('No content. Pass --file PATH, --text "...", or pipe markdown via stdin.');
		process.exit(1);
	}

	const data = await hubApi<HubDriveCreateDocResp>('drive', 'createDoc', {
		name,
		content,
		sourceMime,
		...(flags.account ? { accountSlug: flags.account } : {})
	});
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}
	console.log(`✓ Created Google Doc: ${data.name}`);
	if (data.webViewLink) console.log(`  ${data.webViewLink}`);
}

File: internal/cmd/drive.go
package cmd

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"mm-cli/internal/http"
	"mm-cli/internal/wire"
)

// NewDriveCmd builds the `mm drive` tree.
func NewDriveCmd() *cobra.Command {
	cmd := &cobra.Command{Use: "drive", Short: "Google Drive — list, read, doc-from-markdown, rename/move"}
	cmd.AddCommand(newDriveListCmd(), newDriveReadCmd(), newDriveGetCmd(), newDriveDownloadCmd(), newDriveDocCmd(), newDriveMoveCmd())
	return cmd
}

func newDriveReadCmd() *cobra.Command {
	c := &cobra.Command{
		Use:   "read [id|url]",
		Short: "Read (export) a Google Doc/Sheet/Slide as text",
		Long: "Export a Google-native file's body to text and print it (or write it with --out).\n" +
			"Accepts a bare file id or a pasted Docs/Sheets/Slides URL.\n\n" +
			"Formats: txt (default), html, pdf (use with --out). md is accepted but the\n" +
			"current gateway export path returns 503 for text/markdown — use txt for now.",
		Args: cobra.ExactArgs(1),
		RunE: runDriveRead,
	}
	c.Flags().String("as", "txt", "Export format: txt|html|pdf|md (md needs backend support — see note)")
	c.Flags().String("out", "", "Write to this path instead of stdout")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newDriveGetCmd() *cobra.Command {
	c := &cobra.Command{
		Use:   "get [id|url]",
		Short: "Show file metadata (name, type, size, parents)",
		Args:  cobra.ExactArgs(1),
		RunE:  runDriveGet,
	}
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newDriveDownloadCmd() *cobra.Command {
	c := &cobra.Command{
		Use:   "download [id|url]",
		Short: "Download a non-Google file's raw content (PDF, .md, .csv …)",
		Args:  cobra.ExactArgs(1),
		RunE:  runDriveDownload,
	}
	c.Flags().String("out", "", "Write to this path instead of stdout")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newDriveListCmd() *cobra.Command {
	c := &cobra.Command{Use: "ls", Aliases: []string{"list"}, Short: "List files (filter with --q)", Args: cobra.NoArgs, RunE: runDriveList}
	c.Flags().String("q", "", "Drive search query")
	c.Flags().Int("max", 20, "Max results")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newDriveDocCmd() *cobra.Command {
	c := &cobra.Command{
		Use:   "doc [name]",
		Short: "Create a Google Doc from a markdown file (or stdin)",
		Args:  cobra.MinimumNArgs(1),
		RunE:  runDriveDoc,
	}
	c.Flags().String("file", "", "Path to markdown file (or pipe via stdin)")
	c.Flags().String("text", "", "Inline content")
	c.Flags().String("mime", "text/markdown", "Source mime: text/markdown|text/plain|text/html")
	c.Flags().String("folder", "", "Target folder id")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newDriveMoveCmd() *cobra.Command {
	c := &cobra.Command{
		Use:     "mv [id]",
		Aliases: []string{"rename", "move"},
		Short:   "Rename and/or move a file",
		Args:    cobra.ExactArgs(1),
		RunE:    runDriveMove,
	}
	c.Flags().String("name", "", "New name")
	c.Flags().String("parent", "", "Add a parent (move into)")
	c.Flags().String("unparent", "", "Remove a parent (move out of)")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func runDriveList(cmd *cobra.Command, _ []string) error {
	q, _ := cmd.Flags().GetString("q")
	max, _ := cmd.Flags().GetInt("max")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")
	req := map[string]any{}
	if q != "" {
		req["q"] = q
	}
	if max != 20 && max > 0 {
		req["max"] = max
	}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubDriveListResp
	if err := client.Hub(cmd.Context(), "drive", "list", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	if len(resp.Files) == 0 {
		fmt.Println("No files match.")
		return nil
	}
	for _, f := range resp.Files {
		when := fmtWhen(f.ModifiedTime)
		kind := mimeLabel(f.MimeType)
		link := f.WebViewLink
		if link != "" {
			fmt.Printf("- **%s** | `%s` | [%s](%s)\n", when, kind, f.Name, link)
		} else {
			fmt.Printf("- **%s** | `%s` | %s\n", when, kind, f.Name)
		}
	}
	return nil
}

func runDriveDoc(cmd *cobra.Command, args []string) error {
	name := strings.TrimSpace(strings.Join(args, " "))
	if name == "" {
		return fmt.Errorf("name is required")
	}
	file, _ := cmd.Flags().GetString("file")
	text, _ := cmd.Flags().GetString("text")
	mime, _ := cmd.Flags().GetString("mime")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	var content string
	switch {
	case file != "":
		b, err := os.ReadFile(file)
		if err != nil {
			return fmt.Errorf("Couldn't read --file %s: %w", file, err)
		}
		content = string(b)
	case text != "":
		content = text
	default:
		// Try stdin.
		fi, _ := os.Stdin.Stat()
		if (fi.Mode() & os.ModeCharDevice) == 0 {
			b, err := io.ReadAll(os.Stdin)
			if err != nil {
				return err
			}
			content = string(b)
		} else {
			return fmt.Errorf("No content. Pass --file PATH, --text \"...\", or pipe markdown via stdin.")
		}
	}

	req := map[string]any{
		"name":       name,
		"content":    content,
		"sourceMime": mime,
	}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubDriveCreateDocResp
	if err := client.Hub(cmd.Context(), "drive", "createDoc", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Printf("✓ Created Google Doc: %s\n", resp.Name)
	if resp.WebViewLink != nil && *resp.WebViewLink != "" {
		fmt.Printf("  %s\n", *resp.WebViewLink)
	}
	return nil
}

func runDriveRead(cmd *cobra.Command, args []string) error {
	id := driveFileID(args[0])
	asFlag, _ := cmd.Flags().GetString("as")
	outPath, _ := cmd.Flags().GetString("out")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	mime, err := mimeForAs(asFlag)
	if err != nil {
		return err
	}
	req := map[string]any{"fileId": id, "mimeType": mime}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubDriveExportResp
	if err := client.Hub(cmd.Context(), "drive", "export", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	if outPath != "" {
		if err := os.WriteFile(outPath, []byte(resp.Content), 0o644); err != nil {
			return fmt.Errorf("Couldn't write %s: %w", outPath, err)
		}
		fmt.Fprintf(os.Stderr, "✓ Wrote %s (%s)\n", outPath, resp.MimeType)
		return nil
	}
	fmt.Print(resp.Content)
	return nil
}

func runDriveGet(cmd *cobra.Command, args []string) error {
	id := driveFileID(args[0])
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	req := map[string]any{
		"fileId": id,
		"fields": "id,name,mimeType,modifiedTime,webViewLink,parents,size",
	}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubDriveGetResp
	if err := client.Hub(cmd.Context(), "drive", "get", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Printf("%s\n", resp.Name)
	fmt.Printf("  type:     %s\n", mimeLabel(resp.MimeType))
	if resp.Size != "" {
		fmt.Printf("  size:     %s\n", resp.Size)
	}
	if resp.ModifiedTime != "" {
		fmt.Printf("  modified: %s\n", fmtWhen(resp.ModifiedTime))
	}
	if resp.WebViewLink != "" {
		fmt.Printf("  link:     %s\n", resp.WebViewLink)
	}
	if len(resp.Parents) > 0 {
		fmt.Printf("  parents:  %s\n", strings.Join(resp.Parents, ", "))
	}
	return nil
}

func runDriveDownload(cmd *cobra.Command, args []string) error {
	id := driveFileID(args[0])
	outPath, _ := cmd.Flags().GetString("out")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	req := map[string]any{"fileId": id}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubDriveDownloadResp
	if err := client.Hub(cmd.Context(), "drive", "download", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	if outPath != "" {
		if err := os.WriteFile(outPath, []byte(resp.Content), 0o644); err != nil {
			return fmt.Errorf("Couldn't write %s: %w", outPath, err)
		}
		fmt.Fprintf(os.Stderr, "✓ Wrote %s\n", outPath)
		return nil
	}
	fmt.Print(resp.Content)
	return nil
}

func runDriveMove(cmd *cobra.Command, args []string) error {
	id := args[0]
	name, _ := cmd.Flags().GetString("name")
	parent, _ := cmd.Flags().GetString("parent")
	unparent, _ := cmd.Flags().GetString("unparent")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	req := map[string]any{"fileId": id}
	if name != "" {
		req["name"] = name
	}
	if parent != "" {
		req["addParents"] = []string{parent}
	}
	if unparent != "" {
		req["removeParents"] = []string{unparent}
	}
	if account != "" {
		req["accountSlug"] = account
	}
	if name == "" && parent == "" && unparent == "" {
		return fmt.Errorf("Nothing to do — pass at least one of --name, --parent, --unparent.")
	}

	client := http.New()
	var resp wire.HubDriveUpdateResp
	if err := client.Hub(cmd.Context(), "drive", "update", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Printf("✓ %s\n", resp.Name)
	if resp.WebViewLink != nil && *resp.WebViewLink != "" {
		fmt.Printf("  %s\n", *resp.WebViewLink)
	}
	return nil
}

func fmtWhen(iso string) string {
	if iso == "" {
		return "—"
	}
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return iso
	}
	diff := time.Since(t)
	switch {
	case diff < time.Minute:
		return "just now"
	case diff < time.Hour:
		return fmt.Sprintf("%dm", int(diff.Minutes()+0.5))
	case diff < 24*time.Hour:
		return fmt.Sprintf("%dh", int(diff.Hours()+0.5))
	case diff < 7*24*time.Hour:
		return fmt.Sprintf("%dd", int(diff.Hours()/24+0.5))
	}
	return t.Format("2 Jan")
}

func mimeLabel(m string) string {
	switch {
	case m == "application/vnd.google-apps.document":
		return "doc"
	case m == "application/vnd.google-apps.spreadsheet":
		return "sheet"
	case m == "application/vnd.google-apps.presentation":
		return "slide"
	case m == "application/vnd.google-apps.folder":
		return "folder"
	case strings.HasPrefix(m, "image/"):
		return "img"
	case m == "application/pdf":
		return "pdf"
	}
	if idx := strings.LastIndex(m, "/"); idx >= 0 {
		return m[idx+1:]
	}
	return m
}

// mimeForAs maps the --as shorthand to an export mime type.
func mimeForAs(as string) (string, error) {
	switch strings.ToLower(strings.TrimSpace(as)) {
	case "", "md", "markdown":
		return "text/markdown", nil
	case "txt", "text", "plain":
		return "text/plain", nil
	case "html":
		return "text/html", nil
	case "pdf":
		return "application/pdf", nil
	}
	// Escape hatch: accept a full mime literal (e.g. an office export mime).
	if strings.Contains(as, "/") {
		return as, nil
	}
	return "", fmt.Errorf("unknown --as %q (use md|txt|html|pdf, or a full mime type)", as)
}

// driveFileID accepts a bare id or a pasted Drive/Docs URL and returns the id.
// e.g. https://docs.google.com/document/d/<id>/edit → <id>
func driveFileID(s string) string {
	s = strings.TrimSpace(s)
	if !strings.Contains(s, "/") {
		return s
	}
	// .../d/<id>/... form (Docs, Sheets, Slides, Drive file links)
	if i := strings.Index(s, "/d/"); i >= 0 {
		rest := s[i+3:]
		if j := strings.IndexByte(rest, '/'); j >= 0 {
			rest = rest[:j]
		}
		if j := strings.IndexAny(rest, "?#"); j >= 0 {
			rest = rest[:j]
		}
		return rest
	}
	// ...?id=<id> form (open?id=, uc?id=)
	if i := strings.Index(s, "id="); i >= 0 {
		rest := s[i+3:]
		if j := strings.IndexAny(rest, "&#"); j >= 0 {
			rest = rest[:j]
		}
		return rest
	}
	return s
}
